plugins/repository-gcs/src/main/java/org/opensearch/repositories/gcs/GoogleCloudStorageClientSettings.java

```diff
@@ -202,10 +265,39 @@ static GoogleCloudStorageClientSettings getClientSettings(final Settings setting
             getConfigValue(settings, clientName, CONNECT_TIMEOUT_SETTING),
             getConfigValue(settings, clientName, READ_TIMEOUT_SETTING),
             getConfigValue(settings, clientName, APPLICATION_NAME_SETTING),
-            getConfigValue(settings, clientName, TOKEN_URI_SETTING)
+            getConfigValue(settings, clientName, TOKEN_URI_SETTING),
+            validateAndCreateProxySettings(settings, clientName)
         );
     }
 
+    static ProxySettings validateAndCreateProxySettings(final Settings settings, final String clientName) {
+        final Proxy.Type proxyType = getConfigValue(settings, clientName, PROXY_TYPE_SETTING);
+        final String proxyHost = getConfigValue(settings, clientName, PROXY_HOST_SETTING);
+        final int proxyPort = getConfigValue(settings, clientName, PROXY_PORT_SETTING);
+        final SecureString proxyUserName = getConfigValue(settings, clientName, PROXY_USERNAME_SETTING);
+        final SecureString proxyPassword = getConfigValue(settings, clientName, PROXY_PASSWORD_SETTING);
+        // Validate proxy settings
+        if (proxyType == Proxy.Type.DIRECT
+            && (proxyPort != 0 || Strings.hasText(proxyHost) || Strings.hasText(proxyUserName) || Strings.hasText(proxyPassword))) {
+            throw new SettingsException(
+                "Google Cloud Storage proxy port or host or username or password have been set but proxy type is not defined."
+            );
+        }
+        if (proxyType != Proxy.Type.DIRECT && (proxyPort == 0 || Strings.isEmpty(proxyHost))) {
+            throw new SettingsException("Google Cloud Storage proxy type has been set but proxy host or port is not defined.");
+        }
```

Comment:
Minor - given that this is identical to the logic introduced in #2098 , it might be worth it to move this to a common `StorageSettings` superclass and have the two classes inherit from this. We could also move `getConfigValue` into this parent class.

Reply:
Yes and this is a problem. All these plugins do not have a common repo or something and as a result, all of them contains the same classes e.g. `AccessSocket` is almost the same and with the same functionality for all plugins. I was planning to open a discussion here about improvements for Cloud Plugins. And how we can help with it.

Reply:
Let's have that discussion in an issue so we can close out this PR. I've opened #2155 around it.

Reply:
sure. will leave comments.